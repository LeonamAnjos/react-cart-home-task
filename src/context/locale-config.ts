export type LocaleType = "de-DE" | "en-US";

type LocaleConfig = {
  currency: string;
};

export const localeMap: Record<LocaleType, LocaleConfig> = {
  "de-DE": { currency: "EUR" },
  "en-US": { currency: "USD" },
};
