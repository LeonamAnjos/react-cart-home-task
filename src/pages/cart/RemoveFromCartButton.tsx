import ElementButton from "../../components/Button";
import RemoveShoppingCart from "../../icons/RemoveShoppingCart";

type RemoveFromCartButtonProps = {
  onClick?: () => void | undefined;
};

const RemoveFromCartButton = ({ onClick }: RemoveFromCartButtonProps) => {
  return (
    <ElementButton
      variant="contained"
      color="secondary"
      style={{ marginLeft: "Auto" }}
      endIcon={<RemoveShoppingCart />}
      onClick={onClick}
    >
      Remove
    </ElementButton>
  );
};

export default RemoveFromCartButton;
