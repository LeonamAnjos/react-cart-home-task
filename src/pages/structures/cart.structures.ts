import { Product } from "./product.structures";

export type CartItem = {
  product: Product;
  coverage: number;
  price: number;
};
