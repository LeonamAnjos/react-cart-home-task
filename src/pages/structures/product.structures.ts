export type Product = {
  id: number;
  name: string;
  minCoverage: number;
  maxCoverage: number;
  risk: number;
};

export const calcPrice = (coverage: number, risk: number): number =>
  coverage * risk;
