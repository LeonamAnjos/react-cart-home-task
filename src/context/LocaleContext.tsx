import React, { createContext, ReactNode, useMemo, useState } from "react";
import { localeMap, LocaleType } from "./locale-config";

type LocaleState = {
  getLocale: () => LocaleType | undefined;
  setLocale: (locale: LocaleType) => void;
  currencyFormat: (num: number) => string;
};

const LocaleContext = createContext<LocaleState>({
  getLocale: (): LocaleType | undefined => undefined,
  setLocale: (locale: LocaleType): void => {
    /* EMPTY */
  },
  currencyFormat: (num: number): string => "",
});

export interface LocaleProviderProps {
  children?: ReactNode;
  locale: LocaleType;
}

export const LocaleProvider = ({ children, locale }: LocaleProviderProps) => {
  const [value, setValue] = useState(locale);

  const contextValue = useMemo(() => {
    const formatter = new Intl.NumberFormat(value, {
      style: "currency",
      currency: localeMap[value]?.currency,
      minimumFractionDigits: 2,
    });

    return {
      getLocale: () => value,
      setLocale: (locale: LocaleType) => setValue(locale),
      currencyFormat: (num: number): string => formatter.format(num),
    };
  }, [value, setValue]);

  return (
    <LocaleContext.Provider value={contextValue}>
      {children}
    </LocaleContext.Provider>
  );
};

export default LocaleContext;
