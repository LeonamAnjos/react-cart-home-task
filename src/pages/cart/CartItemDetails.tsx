import React from "react";
import ElementCard from "../../components/ElementCard";
import { useCartStore } from "../stores/useCartStore";
import { CartItem } from "../structures/cart.structures";
import RemoveFromCartButton from "./RemoveFromCartButton";
import CoveragePriceCaption from "../commons/CoveragePriceCaption";

type CartItemProps = {
  item: CartItem;
};

const CartItemDetails = ({ item }: CartItemProps) => {
  const { removeItem } = useCartStore();

  const handleRemoveFromCartClick = () => {
    removeItem(item);
  };

  return (
    <ElementCard
      title={item.product.name}
      actions={<RemoveFromCartButton onClick={handleRemoveFromCartClick} />}
    >
      <CoveragePriceCaption coverage={item.coverage} price={item.price} />
    </ElementCard>
  );
};

export default CartItemDetails;
