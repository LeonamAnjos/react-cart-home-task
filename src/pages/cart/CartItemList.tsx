import React from "react";
import { CartItem } from "../structures/cart.structures";
import CartItemDetails from "./CartItemDetails";

type CartItemListProps = {
  items: CartItem[];
};

const CartItemList = ({ items }: CartItemListProps) => {
  return (
    <>
      {items.map((item) => (
        <CartItemDetails key={item.product.id} item={item} />
      ))}
    </>
  );
};

export default CartItemList;
