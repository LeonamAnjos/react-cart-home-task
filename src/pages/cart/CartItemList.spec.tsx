import { render } from "@testing-library/react";
import { LocaleProvider } from "../../context/LocaleContext";
import { CartItem } from "../structures/cart.structures";
import CartItemList from "./CartItemList";

describe("<CartItemList />", () => {
  const customRender = (items: CartItem[]) => {
    return render(
      <LocaleProvider locale="en-US">
        <CartItemList items={items} />
      </LocaleProvider>
    );
  };

  it("SHOULD render WHEN list is empty", () => {
    const { container } = customRender([]);

    expect(container).toMatchSnapshot();
  });

  it("SHOULD render WHEN list is not empty", () => {
    const items: CartItem[] = [
      {
        product: {
          id: 1,
          name: "Test product 1",
          minCoverage: 10,
          maxCoverage: 90,
          risk: 0.3,
        },
        coverage: 80.0,
        price: 24.0,
      },
      {
        product: {
          id: 2,
          name: "Test product 2",
          minCoverage: 20,
          maxCoverage: 80,
          risk: 0.4,
        },
        coverage: 50.0,
        price: 20.0,
      },
    ];

    const { container } = customRender(items);

    expect(container).toMatchSnapshot();
  });
});
