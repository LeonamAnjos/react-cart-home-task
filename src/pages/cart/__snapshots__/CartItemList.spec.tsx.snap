// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartItemList /> SHOULD render WHEN list is empty 1`] = `<div />`;

exports[`<CartItemList /> SHOULD render WHEN list is not empty 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-16cqd0p-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
        >
          Test product 1
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
        >
          <div
            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-bjfanr-MuiTypography-root"
          >
            Coverage: $80.00
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
        >
          <div
            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-bjfanr-MuiTypography-root"
          >
            Price: $24.00
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiCardActions-root css-1rwjz6-MuiCardActions-root"
    >
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-zcbmsk-MuiButtonBase-root-MuiButton-root"
        data-testid="remove-from-cart-button"
        style="margin-left: auto;"
        tabindex="0"
        type="button"
      >
        Remove
        <span
          class="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="RemoveShoppingCartIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22.73 22.73 2.77 2.77 2 2l-.73-.73L0 2.54l4.39 4.39 2.21 4.66-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h7.46l1.38 1.38c-.5.36-.83.95-.83 1.62 0 1.1.89 2 1.99 2 .67 0 1.26-.33 1.62-.84L21.46 24l1.27-1.27zM7.42 15c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h2.36l2 2H7.42zm8.13-2c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H6.54l9.01 9zM7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-16cqd0p-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
        >
          Test product 2
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
        >
          <div
            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-bjfanr-MuiTypography-root"
          >
            Coverage: $50.00
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
        >
          <div
            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-bjfanr-MuiTypography-root"
          >
            Price: $20.00
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiCardActions-root css-1rwjz6-MuiCardActions-root"
    >
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-zcbmsk-MuiButtonBase-root-MuiButton-root"
        data-testid="remove-from-cart-button"
        style="margin-left: auto;"
        tabindex="0"
        type="button"
      >
        Remove
        <span
          class="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="RemoveShoppingCartIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22.73 22.73 2.77 2.77 2 2l-.73-.73L0 2.54l4.39 4.39 2.21 4.66-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h7.46l1.38 1.38c-.5.36-.83.95-.83 1.62 0 1.1.89 2 1.99 2 .67 0 1.26-.33 1.62-.84L21.46 24l1.27-1.27zM7.42 15c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h2.36l2 2H7.42zm8.13-2c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H6.54l9.01 9zM7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
